### DISCUSSION TITLE
Color picking + various things found while trying to implement a custom color picker

### DISCUSSION THREAD

**Discussion by nununoisy** [slint-ui/slint#3044] [Ideas]
I needed a way for a user to specify a color, and noticed that there isn't a standard color picker widget, so I figured I'd try writing [my own][slintpad]. It'll do for now but it's a lot more limited than I would like it to be - unfortunately, I can't seem to find a good way to implement some features without falling back to native code:
- I can do some math and a bit of array indexing to convert the r,g,b ints into hex to output a CSS hex color. However, it'd be nice if there was an easier way to do this, ideally for hex numbers in general and not just for colors (although either would be fine).
- It would be nice if color objects had a few other methods to get color information, like methods to get RGBA values or a hex color string.
- More options to convert primitive data into color objects would be nice. Examples: `Colors.hsv(float, int, int) -> color`, `Colors.hsl(float, int, int) -> color`, `Colors.css(string) -> color`, etc.
- On that note, I was not able to find a way to convert from a hex color string to r,g,b ints. I can't find a way to extract characters or substrings from the hex color, so I can't use a similar approach to convert a hex value to an int. Making strings indexable would allow this to work and makes sense considering it's standard in many other languages.
- If I use `SpinBox`es for the r,g,b component value labels, their value sometimes fails to update when moving the sliders. It seems to occur after editing the inner textbox. The `SpinBox` fluent component has an internal private function to update the inner text value but there doesn't seem to be a way for me to call it or cause it to trigger, so I have to just use labels for now.

[slintpad]: https://slint.dev/releases/1.1.0/editor/?snippet=import+%7B+Button%2C+VerticalBox+%7D+from+%22std-widgets.slint%22%3B%0A%0Acomponent+ColorSlider+inherits+Rectangle+%7B%0A++++in-out+property%3Cfloat%3E+maximum%3A+255%3B%0A++++in-out+property%3Cfloat%3E+minimum%3A+0%3B%0A++++in-out+property%3Cint%3E+value%3B%0A%0A++++in+property%3Ccolor%3E+left-color%3A+%23000%3B%0A++++in+property%3Ccolor%3E+right-color%3A+%23f00%3B%0A%0A++++callback+moved%28%29%3B%0A%0A++++min-height%3A+24px%3B%0A++++min-width%3A+100px%3B%0A++++horizontal-stretch%3A+1%3B%0A++++vertical-stretch%3A+0%3B%0A%0A++++border-radius%3A+root.height%2F2%3B%0A++++background%3A+%40linear-gradient%2890deg%2C+left-color+0%25%2C+right-color+100%25%29%3B%0A++++border-width%3A+1px%3B%0A++++border-color%3A+%23999%3B%0A%0A++++handle+%3A%3D+Rectangle+%7B%0A++++++++width%3A+self.height%3B%0A++++++++height%3A+parent.height%3B%0A++++++++border-width%3A+3px%3B%0A++++++++border-radius%3A+self.height+%2F+2%3B%0A++++++++background%3A+touch.pressed+%3F+%23f8f%3A+touch.has-hover+%3F+%2366f+%3A+%230000ff%3B%0A++++++++border-color%3A+self.background.darker%2830%25%29%3B%0A++++++++x%3A+%28root.width+-+handle.width%29+*+%28root.value+-+root.minimum%29%2F%28root.maximum+-+root.minimum%29%3B%0A++++%7D%0A++++touch+%3A%3D+TouchArea+%7B%0A++++++++property+%3Cfloat%3E+pressed-value%3B%0A++++++++pointer-event%28event%29+%3D%3E+%7B%0A++++++++++++if+%28event.button+%3D%3D+PointerEventButton.left+%26%26+event.kind+%3D%3D+PointerEventKind.down%29+%7B%0A++++++++++++++++self.pressed-value+%3D+root.value%3B%0A++++++++++++%7D%0A++++++++%7D%0A++++++++moved+%3D%3E+%7B%0A++++++++++++if+%28self.enabled+%26%26+self.pressed%29+%7B%0A++++++++++++++++root.value+%3D+max%28root.minimum%2C+min%28root.maximum%2C%0A++++++++++++++++++++self.pressed-value+%2B+%28touch.mouse-x+-+touch.pressed-x%29+*+%28root.maximum+-+root.minimum%29+%2F+%28root.width+-+handle.width%29%29%29%3B%0A++++++++++++++++root.moved%28%29%3B%0A++++++++++++%7D%0A++++++++%7D%0A++++%7D%0A%7D%0A%0Aexport+component+ColorPicker+%7B%0A++++width%3A+350px%3B%0A%0A++++in-out+property%3Cint%3E+r%3A+255%3B%0A++++in-out+property%3Cint%3E+g%3A+255%3B%0A++++in-out+property%3Cint%3E+b%3A+255%3B%0A++++out+property%3Ccolor%3E+value%3A+%23ffffff%3B%0A%0A++++property%3Cstring%3E+p-hex-color%3A+%22%23ffffff%22%3B%0A%0A++++function+to-hex-digit%28n%3A+int%29+-%3E+string+%7B%0A++++++++if+%28n+%3C+10%29+%7B%0A++++++++++++return+n%3B%0A++++++++%7D+else+%7B%0A++++++++++++return+%28%5B%22a%22%2C+%22b%22%2C+%22c%22%2C+%22d%22%2C+%22e%22%2C+%22f%22%5D%29%5Bn+-+10%5D%3B%0A++++++++%7D%0A++++%7D%0A%0A++++function+to-hex%28n%3A+int%29+-%3E+string+%7B%0A++++++++return+to-hex-digit%28Math.floor%28n+%2F+16%29%29+%2B+to-hex-digit%28n+-+%28Math.floor%28n+%2F+16%29+*+16%29%29%3B%0A++++%7D%0A%0A++++function+p-update%28%29+%7B%0A++++++++p-hex-color+%3D+%22%23%22+%2B+to-hex%28r%29+%2B+to-hex%28g%29+%2B+to-hex%28b%29%3B%0A++++++++value+%3D+rgb%28r%2C+g%2C+b%29%3B%0A++++%7D%0A%0A++++VerticalBox+%7B%0A++++++++alignment%3A+start%3B%0A++++++++HorizontalLayout+%7B%0A++++++++++++alignment%3A+stretch%3B%0A++++++++++++Text+%7B%0A++++++++++++++++text%3A+p-hex-color%3B%0A++++++++++++%7D%0A++++++++++++Rectangle+%7B%0A++++++++++++++++background%3A+rgb%28root.r%2C+root.g%2C+root.b%29%3B%0A++++++++++++++++width%3A+50%25%3B%0A++++++++++++++++border-radius%3A+self.height%2F4%3B%0A++++++++++++++++border-width%3A+1px%3B%0A++++++++++++++++border-color%3A+%23999%3B%0A++++++++++++%7D%0A++++++++%7D%0A++++++++HorizontalLayout+%7B%0A++++++++++++alignment%3A+stretch%3B%0A++++++++++++Text+%7B%0A++++++++++++++++vertical-alignment%3A+center%3B%0A++++++++++++++++text%3A+root.r%3B%0A++++++++++++++++width%3A+30px%3B%0A++++++++++++%7D%0A++++++++++++ColorSlider+%7B%0A++++++++++++++++value+%3C%3D%3E+root.r%3B%0A++++++++++++++++left-color%3A+rgb%280%2C+root.g%2C+root.b%29%3B%0A++++++++++++++++right-color%3A+rgb%28255%2C+root.g%2C+root.b%29%3B%0A++++++++++++++++moved+%3D%3E+%7B+root.p-update%28%29%3B+%7D%0A++++++++++++%7D%0A++++++++%7D%0A++++++++HorizontalLayout+%7B%0A++++++++++++alignment%3A+stretch%3B%0A++++++++++++Text+%7B%0A++++++++++++++++vertical-alignment%3A+center%3B%0A++++++++++++++++text%3A+root.g%3B%0A++++++++++++++++width%3A+30px%3B%0A++++++++++++%7D%0A++++++++++++ColorSlider+%7B%0A++++++++++++++++value+%3C%3D%3E+root.g%3B%0A++++++++++++++++left-color%3A+rgb%28root.r%2C+0%2C+root.b%29%3B%0A++++++++++++++++right-color%3A+rgb%28root.r%2C+255%2C+root.b%29%3B%0A++++++++++++++++moved+%3D%3E+%7B+root.p-update%28%29%3B+%7D%0A++++++++++++%7D%0A++++++++%7D%0A++++++++HorizontalLayout+%7B%0A++++++++++++alignment%3A+stretch%3B%0A++++++++++++Text+%7B%0A++++++++++++++++vertical-alignment%3A+center%3B%0A++++++++++++++++text%3A+root.b%3B%0A++++++++++++++++width%3A+30px%3B%0A++++++++++++%7D%0A++++++++++++ColorSlider+%7B%0A++++++++++++++++value+%3C%3D%3E+root.b%3B%0A++++++++++++++++left-color%3A+rgb%28root.r%2C+root.g%2C+0%29%3B%0A++++++++++++++++right-color%3A+rgb%28root.r%2C+root.g%2C+255%29%3B%0A++++++++++++++++moved+%3D%3E+%7B+root.p-update%28%29%3B+%7D%0A++++++++++++%7D%0A++++++++%7D%0A++++%7D%0A%7D%0A

**Comment by ogoffart** [comment]
Yes, we should have more API on the colors.  See issue https://github.com/slint-ui/slint/issues/911
`Colors.hsv`  makes total sense.  I'm not sure about the `Colors.css(string)`


>  I was not able to find a way to convert from a hex color string to r,g,b ints. I can't find a way to extract characters or substrings from the hex color, so I can't use a similar approach to convert a hex value to an int. Making strings indexable would allow this to work and makes sense considering it's standard in many other languages.

We don't want the slint language to become a general purpose programming language. I think this kindof things should be done in native code.


> If I use SpinBoxes for the r,g,b component value labels, their value sometimes fails to update when moving the sliders. It seems to occur after editing the inner textbox. The SpinBox fluent component has an internal private function to update the inner text value but there doesn't seem to be a way for me to call it or cause it to trigger, so I have to just use labels for now.

Could you elaborate on the problem?  Is that a Bug?

**Comment by nununoisy** [comment]
Understandable - CSS color parsing is probably far too complex for most needs (that's what `css-color-parser` is for!). Let me take a step back and explain my reasoning here - given that there doesn't appear to be any good Rust libraries for displaying a native/OS color picker dialog (that I could find, anyway), I think that a color picker widget could be very useful in the standard library. There are many uses for a color picker, like multimedia programs on desktop (exactly what I'm using Slint for) or LED/display configuration on embedded platforms. 

Most color picking GUIs nowadays support input in the form of hex colors (e.g. `#ff9800`): Qt's `QColorDialog`, GTK's `ColorChooserWidget`, and SwiftUI's `ColorPicker` come to mind. Hex colors are incredibly convenient to work with because they're relatively compact, generally human-readable, and easily copy-pasteable (unlike decimal RGB values), so it'd make a lot of sense for a hypothetical standard color picker widget to support hex colors. However, because of the limitations I detailed in my initial comment, it would be impossible to do that without additional language support. With that said, perhaps this API is a bit closer to your intended scope:
- `Colors.hex(string) -> color` accepts a string of 3 hex octets optionally prefixed with a single `#`, parses it as a 24-bit RGB color, and returns it as a `color`.
  - `Colors.hex("#ff0000") -> #ff0000`
  - `Colors.hex("00ff00") -> #00ff00`
  - `Colors.hex("not a color") -> #000000` (or some other placeholder/sentinel color)
- `color.to-hex() -> string` converts a `color` into a 24-bit hex string prefixed by `#`. This is already possible with a bit of math and const array indexing but it would be nice to have a cleaner way of doing it.
  - `Colors.blue.to-hex() -> "#0000ff"`
  - `Colors.rgb(255, 255, 255).to-hex() -> "#ffffff"`
  - `Colors.rgba(255, 255, 255, 128).to-hex() -> "#ffffff"`
I've intentionally omitted HTML shorthand colors (`#639` being RebeccaPurple) and RGBA hex colors (`#ff00ff80` being Fuschia with 0.5 alpha) for the sake of simplicity, but they might be worth considering as well.

As for the notes on the `SpinBox`, I've done a bit of additional experimenting and I've found a couple of bugs in the current implementation; I'll open issues so they can be tracked in the right place.
